frontend/public/components/routes.tsx

```diff
@@ -200,32 +203,57 @@ const RouteListRow: React.SFC<RoutesRowProps> = ({obj: route}) => <ResourceRow o
 </ResourceRow>;
 
-const TLSSettings = props => <span>
-  {!props.tls && 'TLS is not enabled.'}
-  {props.tls && <dl>
-    <dt>Termination Type</dt>
-    <dd>{props.tls.termination}</dd>
-    <dt>Insecure Traffic</dt>
-    <dd>{props.tls.insecureEdgeTerminationPolicy || '-'}</dd>
-    <dt>Certificate</dt>
-    <dd>{props.tls.certificate || '-'}</dd>
-    <dt>Key</dt>
-    <dd>{props.tls.key || '-'}</dd>
-    <dt>CA Certificate</dt>
-    <dd>{props.tls.caCertificate || '-'}</dd>
-    {_.get(props.tls, 'termination') === 'reencrypt' && <span>
-      <dt>Destination CA Cert</dt>
-      <dd>{props.tls.destinationCACertificate || '-'}</dd>
-    </span>}
-  </dl>
+class TLSSettings extends SafetyFirst<TLSDataProps, TLSDataState> {
+  constructor(props) {
+    super(props);
+    this.state = { showPrivateKey: false };
+    this.toggleKey = this.toggleKey.bind(this);
+  }
+
+  toggleKey() {
+    this.setState({ showPrivateKey: !this.state.showPrivateKey });
+  }
+
+  render() {
+    const { showPrivateKey } = this.state;
+    const { tls } = this.props;
+    const visibleKeyValue = showPrivateKey ? tls.key : <MaskedData /> ;
+
+    return !tls ?
+      'TLS is not enabled.' :
+      <dl>
+        <dt>Termination Type</dt>
+        <dd>{tls.termination}</dd>
+        <dt>Insecure Traffic</dt>
+        <dd>{tls.insecureEdgeTerminationPolicy || '-'}</dd>
+        <dt>Certificate</dt>
+        <dd>{tls.certificate ? <CopyToClipboard value={tls.certificate}/> : '-'}</dd>
+        <dt className="co-m-route-tls-reveal__title">Key {tls.key &&
+          <button className="btn btn-link co-m-route-tls-reveal__btn" type="button" onClick={this.toggleKey}>
+            {
+              showPrivateKey
+                ? <React.Fragment><i className="fa fa-eye-slash" aria-hidden="true"></i> Hide</React.Fragment>
+                : <React.Fragment><i className="fa fa-eye" aria-hidden="true"></i> Reveal</React.Fragment>
+            }
+          </button>}
+        </dt>
+        <dd>{tls.key ? <CopyToClipboard value={tls.key} visibleValue={visibleKeyValue}/> : '-'}</dd>
+        <dt>CA Certificate</dt>
+        <dd>{tls.caCertificate ? <CopyToClipboard value={tls.caCertificate}/> : '-'}</dd>
+        {tls.termination === 'reencrypt' && <React.Fragment>
+          <dt>Destination CA Cert</dt>
+          <dd>{tls.destinationCACertificate ? <CopyToClipboard value={tls.destinationCACertificate}/> : '-'}</dd>
+        </React.Fragment>}
+      </dl>;
   }
-</span>;
+}
 
 const calcTrafficPercentage = (weight: number, route: any) => {
-  let totalWeight = 0;
-  totalWeight += route.spec.to.weight;
-  _.each(route.spec.alternateBackends, alternate => totalWeight += alternate.weight);
-  return (weight/totalWeight*100).toFixed(1);
+  let totalWeight = _.reduce(route.spec.alternateBackends, (result, alternate) => {
+    return result += alternate.weight;
+  }, route.spec.to.weight);
+
+  return totalWeight ? (weight/totalWeight*100).toFixed(1) : (weight*100).toFixed(1);
```

Comment:
If totalWeight is 0, we know weight is 0, too. So we can just return 0.

```js
const percentage = totalWeight ? ((weight / totalWeight) * 100) : 0;
// Format to one decimal place.
return percentage.toFixed(1);
```

But if weight is 0, we can return early without calculating since percentage will always be 0. (We might just use `-` in that case since the function is already returning a string and I think it displays better.)

```js
const calcTrafficPercentage = (weight: number, route: any) => {
  if (!weight) {
    return '-';
  }
````

Not new, but it's a little strange that we do some formatting here, but add the `%` below. `toFixed` already converts to a string, so we can't use the result as a number. Might as well add the `%` here.

```js
const percentage = (weight / totalWeight) * 100;
// Format to one decimal place.
return `${percentage.toFixed(1)}%`;
```

(Sorry I should have caught that on previous review.)

Reply:
fixed, thanks Sam